### PR TITLE
Fixes Changing Account Settings > Notifications > Advanced Notifications on android causes discrepancies with web

### PR DESCRIPTION
**Pull request by langleyd**
Fixes https://github.com/vector-im/element-android/issues/3672

This PR changes the existing logic, which continuously mutated the push rule definitions, updating them by trying to generalise the behaviour based on the `PushRuleKind` and handling exceptions to the definitions, overriding using the `RuleId`. The new implementation matches that on web, by having a static definition for each notification type. 
The new [StandardAction](https://github.com/vector-im/element-android/pull/3681/files#diff-5d8701d9800b6e23c2432dd55667562020900853b44f625ac64f4574425efbbaR21) matches that on the [web](https://github.com/matrix-org/matrix-react-sdk/blob/2089127dad915575d4c7d275ba296c48dc34874a/src/notifications/StandardActions.ts#L22). As do the [PushRuleDefinitions](vector/src/main/java/im/vector/app/features/settings/notifications/PushRuleDefinitions.kt) match those on the [web](https://github.com/matrix-org/matrix-react-sdk/blob/2089127dad915575d4c7d275ba296c48dc34874a/src/notifications/VectorPushRulesDefinitions.ts#L83).

I also [updated the SDK](https://github.com/vector-im/element-android/pull/3681/files#diff-1de85cbaf427204212001e62e1eb726b9cb833d0a6b33c96dffcd0e967b2f65dR26) to match the behaviour of the web based on the static rule definitions, whereby `enabled` request is always sent and `actions` request is sent if we are updating the actions(i.e. `enabled==true`).

I tested selecting each radio option, for each setting from android and they all match the requests(/reflect correctly in the ui) as on web.

### CODE REVIEW

**Comment by langleyd** [comment]
This fix also needs to be incorporated into the V2 logic in https://github.com/vector-im/element-android/pull/3673

File: vector/src/main/java/im/vector/app/core/preference/PushRulePreference.kt [hunk 1]
```diff
@@ -130,10 +130,11 @@ class PushRulePreference : VectorPreference {
                 }
             } else {
                 if (NOTIFICATION_OFF_INDEX == index) {
-                    if (safeKind == RuleSetKey.UNDERRIDE || safeRule.ruleId == RuleIds.RULE_ID_SUPPRESS_BOTS_NOTIFICATIONS) {
+                    if (safeKind == RuleSetKey.UNDERRIDE && safeRule.ruleId != RuleIds.RULE_ID_CALL) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/preference/PushRulePreference.kt:133]: I understand that it was not possible for `safeRule.ruleId == RuleIds.RULE_ID_SUPPRESS_BOTS_NOTIFICATIONS` to be true here. But why checking `RULE_ID_CALL` now?

To be honest, this code is really hard to follow, it misses some comment... :/

File: vector/src/main/java/im/vector/app/core/preference/PushRulePreference.kt [hunk 2]
```diff
@@ -142,7 +143,7 @@ class PushRulePreference : VectorPreference {
                                     && safeRule.ruleId != RuleIds.RULE_ID_INVITE_ME
                                     && NOTIFICATION_NOISY_INDEX == index)
 
-                    if (NOTIFICATION_NOISY_INDEX == index) {
+                    if (NOTIFICATION_NOISY_INDEX == index && RuleIds.RULE_ID_ROOM_NOTIF != safeRule.ruleId) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/preference/PushRulePreference.kt:146]: Same remark, should be nice to have some comment to explain why this code is necessary.

**Review by bmarty — APPROVED** [review]
Thanks for the PR.
I've made some comment, and I agree with you, it's not obvious to understand what's happening here...

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/pushrules/rest/PushRule.kt [hunk 3]
```diff
@@ -100,6 +100,13 @@ data class PushRule(
         )
     }
 
+    /**
+     * Get the highlight status. As spec mentions assume false if no tweak present.
+     */
+    fun getHighlight(): Boolean {
+        return (getActions().firstOrNull { it is Action.Highlight } as? Action.Highlight)?.highlight ?: false
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/pushrules/rest/PushRule.kt:107]: You can rewrite to 
`return getActions().filterIsInstance<Action.Highlight>().firstOrNull()?.highlight.orFalse()`

File: vector/src/main/java/im/vector/app/core/preference/PushRulePreference.kt [hunk 4]
```diff
@@ -22,18 +22,23 @@ import android.view.View
...
-import org.matrix.android.sdk.api.pushrules.RuleSetKey
-import org.matrix.android.sdk.api.pushrules.rest.PushRule
-import org.matrix.android.sdk.api.pushrules.rest.PushRuleAndKind
 
 class PushRulePreference : VectorPreference {
 
+    enum class NotificationIndex(val index: Int) {
+        OFF(0),
+        SILENT(1),
+        NOISEY(2);
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/preference/PushRulePreference.kt:31]: Typo?

File: vector/src/main/java/im/vector/app/features/settings/notifications/StandardAction.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.settings.notifications
+
+import org.matrix.android.sdk.api.pushrules.Action
+
+sealed class StandardAction(
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/notifications/StandardAction.kt:21]: Rename to `StandardActions` ?

**Review by bmarty — COMMENTED** [review]
Thanks for the update. Yes I think it's globally better like that.

Some small remarks, else LGTM

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/pushrules/rest/PushRule.kt [hunk 6]
```diff
@@ -104,7 +105,7 @@ data class PushRule(
      * Get the highlight status. As spec mentions assume false if no tweak present.
      */
     fun getHighlight(): Boolean {
-        return (getActions().firstOrNull { it is Action.Highlight } as? Action.Highlight)?.highlight ?: false
+        return (getActions().firstOrNull { it is Action.Highlight } as? Action.Highlight)?.highlight.orFalse()
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/pushrules/rest/PushRule.kt:108]: Why you did not take my whole suggestion?

Comment (langleyd) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/pushrules/rest/PushRule.kt:108]: As sorry, read it too quickly

**Review by bmarty — APPROVED** [review]
Thanks for the update!
